File: servlet/src/com/badgeportal/BadgeCode.java
package com.badgeportal;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Tamper-evident verification codes.
 *
 * A code looks like:   SF-3A9F-B21C-7E04
 *                      ^^ module prefix, then 12 hex digits of a
 *                         SHA-256 digest in three groups of four.
 *
 * The digest is taken over
 *
 *     studentId : moduleId : issuedAtMillis : secret
 *
 * so the code is bound to the exact badge it was issued for. Because
 * the badge row stores studentId, moduleId and issuedAtMillis, the
 * verifier can RECOMPUTE the expected code from the row itself and
 * compare. Anyone who edits the database -- pointing a valid-looking
 * code at a different student, say -- cannot produce a matching digest
 * without the secret, so the mismatch is detected at verification time.
 *
 * NOTE: this exact algorithm is mirrored byte-for-byte in
 * cgi/badgecode.py. The two implementations MUST stay in step, and
 * bench/benchmark.py asserts that they agree.
 */
public final class BadgeCode {

    private BadgeCode() { }

    /** Builds the canonical string that gets hashed. */
    public static String payload(int studentId, int moduleId, long issuedAtMs, String secret) {
        return studentId + ":" + moduleId + ":" + issuedAtMs + ":" + secret;
    }

    /** SHA-256 of the payload, lower-case hex. */
    public static String digestHex(int studentId, int moduleId, long issuedAtMs, String secret) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] out = md.digest(
                payload(studentId, moduleId, issuedAtMs, secret).getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder(out.length * 2);
            for (byte b : out) {
                sb.append(Character.forDigit((b >> 4) & 0xF, 16));
                sb.append(Character.forDigit(b & 0xF, 16));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            // SHA-256 is required of every JVM; this cannot happen.
            throw new IllegalStateException("SHA-256 unavailable", e);
        }
    }

    /**
     * The public verification code for a badge.
     *
     * @param prefix     two-letter module stem, e.g. "SF"
     */
    public static String generate(String prefix, int studentId, int moduleId,
                                  long issuedAtMs, String secret) {
        String hex = digestHex(studentId, moduleId, issuedAtMs, secret)
                        .substring(0, 12).toUpperCase();
        return prefix.toUpperCase() + "-"
             + hex.substring(0, 4) + "-"
             + hex.substring(4, 8) + "-"
             + hex.substring(8, 12);
    }

    /** Badge tier from the module score (stretch goal). */
    public static String tierFor(double score) {
        if (score >= 90.0) return "GOLD";
        if (score >= 75.0) return "SILVER";
        return "BRONZE";
    }

    /**
     * Constant-time-ish comparison of two codes. Verification codes are
     * public values rather than passwords, so this is belt-and-braces,
     * but it costs nothing and keeps the habit right.
     */
    public static boolean matches(String a, String b) {
        if (a == null || b == null || a.length() != b.length()) return false;
        int diff = 0;
        for (int i = 0; i < a.length(); i++) {
            diff |= a.charAt(i) ^ b.charAt(i);
        }
        return diff == 0;
    }

    /** Normalises user input: trims, upper-cases, tolerates missing dashes. */
    public static String normalise(String raw) {
        if (raw == null) return "";
        String s = raw.trim().toUpperCase().replace(" ", "");
        if (s.indexOf('-') < 0 && s.length() == 14) {
            // e.g. SF3A9FB21C7E04 -> SF-3A9F-B21C-7E04
            s = s.substring(0, 2) + "-" + s.substring(2, 6) + "-"
              + s.substring(6, 10) + "-" + s.substring(10, 14);
        }
        return s;
    }
}

File: cgi/badgecode.py
"""
Tamper-evident verification codes -- Python side.

This is a line-for-line mirror of servlet/src/com/badgeportal/BadgeCode.java.
The two MUST agree: the CGI verifier recomputes the digest of a badge
row exactly the way the servlet issuer computed it, and if the two
implementations ever drifted apart, every badge would suddenly look
tampered with when checked through CGI.

bench/benchmark.py guards against that drift by asserting that both
endpoints return the same JSON for the same code.
"""

import hashlib

# Stand-ins used when a badge came from an approved claim rather than a
# course module. A claim badge has no module, so there is no module id
# to hash and no two-letter module prefix to print.
#
# These MUST match BadgeCode.CLAIM_PREFIX and BadgeCode.NO_MODULE on the
# Java side. Change either and every claim-issued badge stops verifying,
# because the digest recomputed here would no longer match the digest
# computed by the issuer.
CLAIM_PREFIX = "CB"
NO_MODULE = 0


def payload(student_id, module_id, issued_at_ms, secret):
    """The canonical string that gets hashed."""
    return "%d:%d:%d:%s" % (student_id, module_id, issued_at_ms, secret)


def digest_hex(student_id, module_id, issued_at_ms, secret):
    """SHA-256 of the payload, lower-case hex."""
    raw = payload(student_id, module_id, issued_at_ms, secret).encode("utf-8")
    return hashlib.sha256(raw).hexdigest()


def generate(prefix, student_id, module_id, issued_at_ms, secret):
    """
    The public verification code for a badge, e.g. SF-3A9F-B21C-7E04.

    Twelve hex digits of the digest, upper-cased, in three groups of
    four, behind the two-letter module prefix.
    """
    hex12 = digest_hex(student_id, module_id, issued_at_ms, secret)[:12].upper()
    return "%s-%s-%s-%s" % (prefix.upper(), hex12[0:4], hex12[4:8], hex12[8:12])


def tier_for(score):
    """Badge tier from the module score."""
    if score >= 90.0:
        return "GOLD"
    if score >= 75.0:
        return "SILVER"
    return "BRONZE"


def matches(a, b):
    """Length-safe comparison of two codes."""
    if a is None or b is None or len(a) != len(b):
        return False
    diff = 0
    for x, y in zip(a, b):
        diff |= ord(x) ^ ord(y)
    return diff == 0


def normalise(raw):
    """Trims, upper-cases, and tolerates a code typed without dashes."""
    if not raw:
        return ""
    s = raw.strip().upper().replace(" ", "")
    if "-" not in s and len(s) == 14:
        s = "%s-%s-%s-%s" % (s[0:2], s[2:6], s[6:10], s[10:14])
    return s
